File: src/create/create.cpp
#include "create.h"

/**
 * Function to handle parameters and transaction code generation
 * for the create command which is responsible for creating new
 * users in the system
 * NOTE: Only usable by Admins
 */

void Create()
{
    if (CheckPermission(AppState::getInstance().getCurrentUser().privilege_type, "create"))
    {
        string username;
        string permission;
        string password;
        int password_encrypted = 0;

        cout << "Please enter new user’s username:" << endl;
        cin.ignore();
        getline(cin, username);

        switch (ValidateUsername(username))
        {
        case 1:
            cout << "Please enter a username that is less than 16 characters long" << endl;
            return;
        case 2:
            cout << "Please enter a username that is at least 1 character long" << endl;
            return;
        case 3:
            cout << "Transaction Failed! Username already exists!" << endl;
            return;
        }

        cout << "Please enter the account type:" << endl;
        getline(cin, permission);

        if (!ValidatePermission(permission))
        {
            cout << "Please enter a valid account type (AA,FS,BS,SS)" << endl;
            return;
        }

        cout << "Please enter the password for the account (must be less than 15 characters long):" << endl;
        getline(cin, password);

        for (int i = 0; i < password.size(); i++){
            password_encrypted += int(password[i]);
        }

        // TODO Still need to implement the code to create the account

        string transaction_code = "01 ";

        transaction_code += username;
        for (int i = 0; i < 16 - username.length(); i++)
        {
            transaction_code += ' ';
        }
        transaction_code += permission + " ";
        transaction_code += "000000.00 ";

        transaction_code +=  to_string(password_encrypted);
        transaction_code += "\n";

        AppState::getInstance().appendTransactionBuffer(transaction_code);
        /*
        CREATE Transaction Format: 

        XX_UUUUUUUUUUUUUUU_TT_CCCCCCCCC_PPPPPPPPPP
        XX - The 2 Digit Transaction Code
        UUUUUUUUUUUUUUU - The Maximum 15 Character Username
        TT - The 2 Digit Account Type Code of the User
        CCCCCCCCC - The Available Credit of the Account
        PPPPPPPPPP - the password for the account
        _ - Represents a Space

        */
        map<string, string> privileges{{"AA", "Admin"}, {"FS", "Full Standard"}, {"BS", "Buy Standard"}, {"SS", "Sell Standard"}};

               cout << "User \"" << username << "\" will be created upon logout with permission type " 
                    << privileges[permission] << " and a balance of $0 credits." << endl;
    }
    else
    {
        cout << "Transaction Failed! You do not have permission to create new accounts!" << endl;
    }
}

File: srcBack/create.py
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# File Name: create.py
#
# Description:
# The following file contains the function used to handle "create" transactions
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

import utility

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Function Name: advertise
#
# Description:
# The following function handles the "create" transaction, it takes in the new user's username
# the account type, starting credit, encrypted password, and the users list storing the information
# from the Users file. It will add this new user's information to the users list to be written
# to the Users file later.
#
# Parameters:
# username - The name of the new account
# account_type - The privilege of the new account
# credit - The credits that the new account will have
# password - The password of the new account
# users - The list of Users storing the information from the Users File
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
def create(username, account_type, credit, password, users):
    # Creates a new user and adds them to the users list
    acc = 0
    no_error = True
    for user in users:
        if (user.username == username):
            no_error = False
        acc += 1
    if no_error == False:
        print('Error: User already exists in Users File!')
        return
    user = utility.User(username, account_type, credit, password)
    users.append(user)